rfcs/20190815-tfx-notebook.md

```diff
@@ -0,0 +1,563 @@
+    warning and return.
+*   Bi-directional import to notebook from pipeline would "just work".
+
+Cons:
+
+*   InteractiveContext is still present in the production pipeline as a no-op /
+    affects readability.
+*   Puts the burden on user to scrub out calls to InteractiveContext.
+
+##### Option 3: Mark lines/cells to be skipped during export.
+
+Add custom magic to mark lines/cells as skip_for_export, can also be used by the
+user to skip scratch work in cells.
+
+Example line magic:
+
+```
+%skip_for_export sess = InteractiveContext()
+...
+%skip_for_export sess.run(example_gen)
+```
+
+Example cell magic:
+
+```
+%%skip_for_export
+# Cell contains scratch work that doesn't need to be exported.
+...
+```
+
+Cons:
+
+*   Puts burden on the user to filter out the InteractiveContext objects. User
+    may forget to mark some usages of InteractiveContext, meaning
+    InteractiveContext instances can get leaked to the final pipeline.
+
+##### Option 4: Delete the lines containing InteractiveContext variables.
+
+Cons:
+
+*   Not robust to duplicate references.
+*   We can find the references to InteractiveContext by either keeping track of
+    them weakly within the class on __init__, or we can use gc module to
+    dynamically find the references. But then finding and deleting all
+    associated lines with each instance seems hard.
+    *   What if user makes a helper function and passes in a context variable?
+        (not likely, but possible)
+
+Note each of these options only filters the InteractiveContext usage in the
+exported python script, and does not prevent the user from adding it back
+afterwards.
+
+#### Export notebook contents to pipeline
+
+1.  Present the user with a Beam pipeline export template cell. Airflow/Kubeflow
+    template code can be linked to in documentation, or populated in additional
+    cells with code commented.
+    1.  User fills out any globals/configuration code specific to
+        Beam/Airflow/Kubeflow.
+    1.  User fills out a `pipeline.Pipeline()` instance to export.
+        1.  We can alternatively have the user wrap the pipeline.Pipeline()
+            instance in a function, like `_create_pipeline(...)` in the existing
+            [pipeline examples](https://github.com/tensorflow/tfx/blob/master/tfx/examples/chicago_taxi_pipeline/taxi_pipeline_simple.py),
+            but this could make pipeline export more cumbersome for users who
+            have not organized their notebook in such a way. We could also
+            potentially update the notebook example to push users into a
+            particular notebook organization.
+1.  When the user runs the cell, or more specifically, when
+    `context.export_to_pipeline()` is executed, export the notebook code to .py
+    file.
+    1.  It seems beneficial to keep the export_to_pipeline() line in the same
+        cell as the pipeline.Pipeline() declaration so the user can fix any
+        errors before the export happens.
```

Comment:
By the nature of notebooks, this is hard to do. Why is it necessary to have these be in the same cell? Should Pipeline() just call export at the end if these are paired processes?

Reply:
I think this comment is a minor point regarding keeping the following calls in the same cell for something like the following, so that they can be modified / executed as one unit.

```python
my_pipeline = pipeline.Pipeline(components=[ExampleGen, Trainer, ...])
context.export_to_pipeline(my_pipeline, 'my_notebook.ipynb', 'exported_pipeline.py')
```

The idea would be that it would be less confusing in a single cell for any Pipeline definitions to always be coupled to the export step.